core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -412,12 +416,69 @@ public final SegmentsStats segmentsStats() {
+                memoryResources.addStoredFieldsInBytes(guardedRamBytesUsed(segmentReader.getFieldsReader()));
+                memoryResources.addTermVectorsInBytes(guardedRamBytesUsed(segmentReader.getTermVectorsReader()));
+                memoryResources.addNormsInBytes(guardedRamBytesUsed(segmentReader.getNormsReader()));
+                memoryResources.addDocValuesInBytes(guardedRamBytesUsed(segmentReader.getDocValuesReader()));
+                stats.addMemoryIndexResources(memoryResources);
+                
+                Directory directory = null;
+                SegmentInfo segmentInfo = segmentReader.getSegmentInfo().info;
+                boolean useCompoundFile = segmentInfo.getUseCompoundFile();
+                if (useCompoundFile) {
+                    try {
+                        directory = engineConfig.getCodec().compoundFormat().getCompoundReader(segmentReader.directory(), segmentInfo, IOContext.READ);
+                    } catch (IOException e) {
+                        logger.trace("error when opening compound reader for segment [{}]", e, segmentInfo.name);
+                    }
+                } else {
+                    directory = segmentReader.directory();
+                }
+                
+                // Couldn't open compound reader, skip
+                if (directory == null) {
+                    continue;
+                }
+                
+                try {
+                    String[] files = directory.listAll();
+                    for (String file : files) {
+                        String extension = IndexFileNames.getExtension(file);
+                        if (extension == null) {
+                            continue;
+                        }
+                        
+                        long length = directory.fileLength(file);
+                        
+                        if ("tip".equals(extension) || "tim".equals(extension)
+                                || "pos".equals(extension) || "doc".equals(extension) || "pay".equals(extension)) {
+                            diskResources.addTermsInBytes(length);
+                        } else if ("fdt".equals(extension) || "fdx".equals(extension)) {
+                            diskResources.addStoredFieldsInBytes(length);
+                        } else if ("dvd".equals(extension) || "dvm".equals(extension)) {
+                            diskResources.addDocValuesInBytes(length);
+                        } else if ("nvd".equals(extension) || "nvm".equals(extension)) {
+                            diskResources.addNormsInBytes(length);
+                        } else if ("tvd".equals(extension) || "tvx".equals(extension) || "tvf".equals(extension)) {
+                            diskResources.addTermVectorsInBytes(length);
+                        }
+                    }
+                } catch (IOException e) {
+                    // TODO maybe catch specific exceptions for gone files?
```

Comment:
It's expected that some files will sometimes be gone here, because `IndexWriter` is actively changing the index while this reader has a point-in-time snapshot open.  Maybe we catch `FileNotFoundException` and `NoSuchFileException` and don't log (even trace) those ones?  Any other `IOException` we should log.